### PR TITLE
added AvailableIpAddresses in helper

### PR DESCRIPTION
**Pull request by JasonRuonanWang**
added a helper function for getting all available IP addresses. This could be useful for all socket based engines and transports.

### CODE REVIEW

File: source/adios2/helper/adiosSystem.cpp [hunk 1]
```diff
@@ -74,6 +80,45 @@ bool IsZeroIndexed(const std::string hostLanguage) noexcept
     return isZeroIndexed;
 }
 
+std::vector<std::string> AvailableIpAddresses() noexcept
```
Comment (williamfgc) [inline source/adios2/helper/adiosSystem.cpp]:
```suggestion
#ifndef _WIN32
std::vector<std::string> AvailableIpAddresses() noexcept
```

File: source/adios2/helper/adiosSystem.cpp [hunk 2]
```diff
@@ -74,6 +80,45 @@ bool IsZeroIndexed(const std::string hostLanguage) noexcept
...
+        {
+            ips.emplace_back(ip);
+        }
+        ++p;
+    }
+    if_freenameindex(head);
+    close(socket_handler);
+    return ips;
+}
+
```
Comment (williamfgc) [inline source/adios2/helper/adiosSystem.cpp]:
```suggestion
#endif
```

File: source/adios2/helper/adiosSystem.h [hunk 3]
```diff
@@ -57,6 +57,12 @@ bool IsRowMajor(const std::string hostLanguage) noexcept;
  */
 bool IsZeroIndexed(const std::string hostLanguage) noexcept;
 
+/**
+ * returns a vector of strings with all available IP addresses on the node
+ * @return vector of strings
+ */
```
Comment (williamfgc) [inline source/adios2/helper/adiosSystem.h]:
```suggestion
 */
 #ifndef _WIN32
```

File: source/adios2/helper/adiosSystem.h [hunk 4]
```diff
@@ -57,6 +57,12 @@ bool IsRowMajor(const std::string hostLanguage) noexcept;
...
+std::vector<std::string> AvailableIpAddresses() noexcept;
+
```
Comment (williamfgc) [inline source/adios2/helper/adiosSystem.h]:
```suggestion
#endif
```

File: source/adios2/helper/adiosSystem.cpp [hunk 5]
```diff
@@ -9,6 +9,12 @@
  */
 #include "adiosSystem.h"
```
Comment (williamfgc) [inline source/adios2/helper/adiosSystem.cpp]:
```suggestion
#ifndef _WIN32
```

File: source/adios2/helper/adiosSystem.cpp [hunk 6]
```diff
@@ -9,6 +9,12 @@
...
 
+#include <arpa/inet.h> //AvailableIpAddresses() inet_ntoa
+#include <net/if.h>    //AvailableIpAddresses() struct if_nameindex
+#include <string.h>    //AvailableIpAddresses() strncp
+#include <sys/ioctl.h> //AvailableIpAddresses() ioctl
+#include <unistd.h>    //AvailableIpAddresses() close
+
```
Comment (williamfgc) [inline source/adios2/helper/adiosSystem.cpp]:
```suggestion
#endif
```

File: source/adios2/helper/adiosSystem.cpp [hunk 7]
```diff
@@ -74,6 +80,45 @@ bool IsZeroIndexed(const std::string hostLanguage) noexcept
...
+        {
+            if (errno == EADDRNOTAVAIL)
+            {
+                ++p;
+                continue;
+            }
+            close(socket_handler);
+            return ips;
+        }
+        std::string ip =
```
Comment (williamfgc) [inline source/adios2/helper/adiosSystem.cpp]:
```suggestion
        const std::string ip =
```

**Comment by JasonRuonanWang** [comment]
@williamfgc Thanks for the suggestions. But there is no way to take them all in a single commit. I tried by they went into 7 separate commits. Shall I re-submit another PR?

**Comment by williamfgc** [comment]
@JasonRuonanWang sure, whatever works better for you. The suggestions button is on beta at github.

**Pull request merged**